packages/runtime-core/src/renderer.ts

```diff
@@ -2119,8 +2119,9 @@ function baseCreateRenderer(
       } else if (
         dynamicChildren &&
         // #1153: fast path should not be taken for non-stable (v-for) fragments
-        (type !== Fragment ||
-          (patchFlag > 0 && patchFlag & PatchFlags.STABLE_FRAGMENT))
+        type !== Fragment &&
```

Comment:
This is not the proper fix. 
the root cause is: `NEED_PATCH` is missing when an element has both `key` and `ref`.
note:
- if the element has a `ref`, it should be marked `NEED_PATCH`
- if the element has a `key`, it should be a block. (shouldUseBlock=true)

https://github.com/vuejs/core/blob/b775b71c788499ec7ee58bc2cf4cd04ed388e072/packages/compiler-core/src/transforms/transformElement.ts#L749-L755

Reply:
The constType in vfor is modified. This happens when literals are used for traversal. Is this the problem here?

https://github.com/vuejs/core/blob/b775b71c788499ec7ee58bc2cf4cd04ed388e072/packages/compiler-core/src/transforms/transformExpression.ts#L351

Reply:
No. I feel it should be: `!(shouldUseBlock && !hasRef) &&` (line:750)

Reply:
Thanks for your advice, it has been modified.